librbd: optionally unregister "laggy" journal clients

trociny opened the pull request:
Fixes: http://tracker.ceph.com/issues/14738

dillaman commented:
@trociny Instead of unregistering the laggy client (which would work), I put a placeholder ```ClientState``` in the registration record [1]. My original idea was to set this to DISCONNECTED (via a new cls method) and adjust the JournalTrimmer to ignore clients with the state set to DISCONNECTED when determining the minimum set. 

The rbd-mirror daemon can then detect if and when it's client state is set to DISCONNECTED, stop replay (if in-progress), and re-initiate an image sync w/o the need to split-brain.  The sync point record supports a "from snapshot" so it can be modified to only resync from the last known good snapshot (if any). For example, if the system gets disconnected during the initial sync, it can recover w/o starting over from scratch by reconnecting to the journal, creating a second sync point record tied to the first, and once the first sync is complete, it would start with the second (similar to how VM/storage live migration transfers the full base, and then one or more deltas until it can cut over).

The iterative / recoverable image sync wouldn't be part of this effort, but I wanted to lay out my full plans.

[1] https://github.com/ceph/ceph/blob/master/src/cls/journal/cls_journal_types.h#L79

trociny commented:
@dillaman Got it! Thanks. Will redo.

trociny commented:
@dillaman How do you like this version?

dillaman commented on src/common/config_opts.h:
```diff
@@ -1250,6 +1250,7 @@ OPTION(rbd_journal_object_flush_interval, OPT_INT, 0) // maximum number of pendi
 OPTION(rbd_journal_object_flush_bytes, OPT_INT, 0) // maximum number of pending bytes per journal object
 OPTION(rbd_journal_object_flush_age, OPT_DOUBLE, 0) // maximum age (in seconds) for pending commits
 OPTION(rbd_journal_pool, OPT_STR, "") // pool for journal objects
+OPTION(rbd_journal_client_object_sets_behind_max, OPT_INT, 0) // maximum number of object sets a journal client can be behind before it is automatically unregistered
```
Minor: perhaps ```rbd_journal_max_concurrent_object_sets```?

dillaman commented on src/journal/JournalMetadata.cc:
```diff
@@ -748,7 +751,13 @@ void JournalMetadata::handle_refresh_complete(C_Refresh *refresh, int r) {
+      r = -ENOENT;
+    } else if (it->state == cls::journal::CLIENT_STATE_DISCONNECTED) {
+      lderr(m_cct) << "client flagged disconnected: " << m_client_id << dendl;
+      r = -ENOTCONN;
```
Even if the client isn't connected, I think you would still want to initialize the journaler. rbd-mirror, during bootstrap can check its connected and use a helper method to re-connect if disconnected and within ImageReplayer it should have a listener for journal updates to check if it has been disconnected while replaying.

dillaman commented on src/journal/JournalMetadata.cc:
```diff
@@ -985,6 +991,8 @@ void JournalMetadata::committed(uint64_t commit_tid,
     ldout(m_cct, 20) << "updated commit position: " << commit_position << ", "
                      << "on_safe=" << m_commit_position_ctx << dendl;
+
+    disconnect_laggy_clients();
```
I would only run this logic as part of the scheduled commit task.  The JournalTrimmer watches for header updates before it proceeds with the trim, so if it's part of the scheduled commit task, it could update the commit position, disconnect the laggy peers, and throw the update event for the Trimmer to detect the newly disconnected client.

dillaman commented on src/journal/JournalMetadata.cc:
```diff
@@ -1022,6 +1030,79 @@ void JournalMetadata::handle_notified(int r) {
+				    cls::journal::CLIENT_STATE_DISCONNECTED);
+
+	librados::AioCompletion *comp =
+	librados::Rados::aio_create_completion(ctx, nullptr,
```
```C_NotifyUpdate``` could be re-used multiple times.

dillaman commented on src/journal/JournalMetadata.cc:
```diff
@@ -1022,6 +1030,79 @@ void JournalMetadata::handle_notified(int r) {
+    clients.pop();
+  }
+
+  assert(clients.size() == 1);
```
Why the loop and then this extra, duplicate code block?

trociny replied:
Because in the loop, I am building the code into the context for all clients but one, while below I execute the code for the last client passing the built context.

Do you suggest just to build the context for all clients, and then run `ctx->complete(0)`?

dillaman replied:
Oh -- I missed the FunctionContext.  Maybe move the bulk librados code section to a helper method so you can re-use it.  You also should be able to just keep adding ```client::client_update_state``` ops to a single ```librados::ObjectWriteOperation``` since each will update different omap keys.

dillaman commented on src/journal/JournalMetadata.cc:
```diff
@@ -1022,6 +1030,79 @@ void JournalMetadata::handle_notified(int r) {
+
+  std::queue<std::string> clients;
+  for (auto &c : m_registered_clients) {
+    if (c.id == m_client_id ||
```
I think we need a configurable filter to exclude certain peers from being disconnected.  We would never want to disconnect the ```librbd::Journal::IMAGE_CLIENT_ID```, only other clients.

dillaman commented on src/tools/rbd_mirror/image_replayer/BootstrapRequest.cc:
```diff
@@ -194,6 +194,10 @@ void BootstrapRequest<I>::handle_get_client(int r) {
     return;
+  } else if (m_client.state == cls::journal::CLIENT_STATE_DISCONNECTED) {
+    dout(0) << ": client flagged disconnected, re-registering" << dendl;
+    unregister_client();
```
So long as this doesn't result in a split-brain, this is probably fine for short term. I think the intermediate term goal would be to set the ```MirrorPeerState``` state back to ```MIRROR_PEER_STATE_SYNCING``` to indicate that a full re-sync is required and re-connect the client. The long term goal would add the smarts to have multiple sync points / incremental syncs upon disconnect.

dillaman replied:
I think you can just fail with some special result code that ```ImageReplayer``` can use to know that it was disconnected and should stop (with error).

trociny replied:
So, do we want that after disconnect the mirror does not reconnect automatically?

Right now, if "disconnected" is detected during replaying, the image repayer stops (without "manual" flag), so it is restarted in max 30 second interval by pool Replayer. And in bootstrap we reregister journal client, so the restart succeeds.

Do you mean the bootstap always returns error here and the "start" is terminated, similarly to "remote image is non-primary or local image is primary" case? So a user intervention is required to manually start image replay. And then how user is supposed to do this? Don't we want an automatic restart, may be after some configurable time interval?

dillaman replied:
Instead of unregister / re-register, why not just update the connected state back to connected (and have the cls method clear the current commit position). How about a new config value for whether or not to automatically resync after a disconnect? Until we get the delta image sync logic, I'd hate to have an overloaded system get behind on replay and then just automatically start a full resync.

dillaman commented:
@trociny Looking good to me.  The only missing piece is the smarts in ```rbd::mirror::ImageReplayer``` to catch a disconnect while we are actively replaying.

trociny commented:
@dillaman Updated, still not tested well. Also there are potential issues to discuss:

- A laggy client is disconnected, resync is started, the client is re-registered. While it is resyncing the master commit position grows and the mirror client is unregistered again before resync is complete. Should we add a way the rbd-mirror could tell that the client cannot be disconnected (during  cync/bootstrap)?

- What to do when disconnect is detected when replaying (or just after resync)? Doing resync immediately does not look like a good idea (mirroring is likely overloaded, and resync will only increase the load). Currently it just stops image replay (though without manual flag, so will be restarted on the next pool rescan). May be schedule resync after configurable interval?

trociny commented:
Added some tests.  Also, experimental `rbd journal disconnect` command is added. It is used in functional tests, not sure though it is useful enough to add it upstream. @dillaman do you think it would be useful  to have such command? It could be also implemented as `rbd mirror image disconnect`.

dillaman commented on src/journal/JournalMetadata.cc:
```diff
@@ -1023,6 +1031,60 @@ void JournalMetadata::handle_notified(int r) {
+    const std::string &client_id = c.id;
+    uint64_t object_set = 0;
+    if (!c.commit_position.object_positions.empty()) {
+      auto position = *(c.commit_position.object_positions.begin());
```
Minor: ```auto &```

dillaman commented on src/test/journal/RadosTestFixture.cc:
```diff
@@ -68,10 +68,12 @@ int RadosTestFixture::create(const std::string &oid, uint8_t order,
   journal::Settings settings;
   settings.commit_interval = commit_interval;
   settings.max_fetch_bytes = max_fetch_bytes;
+  settings.max_concurrent_object_sets= max_concurrent_object_sets;
```
Minor: space before ```=```.

dillaman commented on src/tools/rbd_mirror/ImageReplayer.cc:
```diff
@@ -500,6 +510,26 @@ void ImageReplayer<I>::handle_init_remote_journaler(int r) {
+
+  dout(20) << "client=" << client << dendl;
+
+  if (client.state != cls::journal::CLIENT_STATE_CONNECTED) {
```
I would put this logic into the bootstrap state machine since it already pulls the remote peer record -- and eventually it will support auto-resyncing.

trociny replied:
@dillaman  Do you mean "check if it is disconnected" logic? It is here, after add_listener, just to recheck that the client was not disconnected in time window between we checked the state in bootstrap and registered the listener (so we lost the notification). Or do you want the listener is registered in bootstrap?

dillaman replied:
Missed that -- I thought that the remote journal was initialized first and then bootstrap was invoked.

dillaman commented on src/tools/rbd_mirror/ImageReplayer.cc:
```diff
@@ -1437,6 +1468,25 @@ void ImageReplayer<I>::handle_shut_down(int r, Context *on_start) {
+
+  if (client.state != cls::journal::CLIENT_STATE_CONNECTED) {
+    dout(0) << "client flagged disconnected, stopping image replay" << dendl;
+    stop(); // XXX: Schedule re-sync after configurable interval?
```
Should this stop with an error code so that ``` rbd mirror image status``` reports that it has been disconnected?

dillaman commented on src/tools/rbd/action/Journal.cc:
```diff
@@ -973,6 +1062,10 @@ Shell::Action action_reset(
   &get_reset_arguments, &execute_reset);
 
+Shell::Action action_disconnect(
+  {"journal", "disconnect"}, {}, "Flag image journal client disconnected.", "",
```
How about "journal client disconnect"? Since I can also see the debugging need for "journal client remove"

trociny commented on src/journal/JournalMetadata.cc:
```diff
@@ -986,6 +993,7 @@ void JournalMetadata::committed(uint64_t commit_tid,
 
     ldout(m_cct, 20) << "updated commit position: " << commit_position << ", "
                      << "on_safe=" << m_commit_position_ctx << dendl;
+
```
TODO: remove empty line

trociny commented:
@dillaman updated. Now when a journal client is disconnected, rbd-mirror just stops image replayer, until 'mirror image resync' command is issued by user. Also rbd_mirror_resync_after_disconnect configuration option is added to automatically start the resync.

dillaman commented on src/tools/rbd/action/Journal.cc:
```diff
@@ -985,6 +1084,11 @@ Shell::Action action_import(
 
+Shell::Action action_disconnect(
+  {"journal", "client", "disconnect"}, {},
+  "Flag image journal client disconnected.", "",
```
Minor: "... as disconnected"

dillaman commented on src/tools/rbd_mirror/ImageReplayer.cc:
```diff
@@ -503,6 +513,23 @@ void ImageReplayer<I>::handle_init_remote_journaler(int r) {
+
+  if (client.state != cls::journal::CLIENT_STATE_CONNECTED) {
+    dout(5) << "client flagged disconnected, stopping image replay" << dendl;
+    on_start_fail(0, "disconnected");
```
Should this set an error code so that "mirror image status" shows an error?

dillaman commented on src/tools/rbd_mirror/ImageReplayer.cc:
```diff
@@ -1442,6 +1473,23 @@ void ImageReplayer<I>::handle_shut_down(int r) {
+
+  if (client.state != cls::journal::CLIENT_STATE_CONNECTED) {
+    dout(0) << "client flagged disconnected, stopping image replay" << dendl;
+    stop(nullptr, false, "disconnected");
```
Same comment here: should we set an error code so that an admin can see the problem?

dillaman commented on src/librbd/ImageCtx.cc:
```diff
@@ -930,7 +930,8 @@ struct C_InvalidateCache : public Context {
         "rbd_journal_max_payload_bytes", false)(
-        "rbd_journal_max_concurrent_object_sets", false);
+        "rbd_journal_max_concurrent_object_sets", false)(
+        "rbd_mirror_resync_after_disconnect", false);
```
No need to treat this as a librbd image option -- just use it from rbd-mirror daemon only

trociny replied:
It is not that I insists, though I can imagine a situation when you want the most of the images not to resync after disconnect (in order not to increase the load) while still have a group of "important" images you want to resync ASAP. Setting this option for these images via image-meta makes this possible. Also it simplifies testing -- no need to restart daemon to apply the option change.

@dillaman If you think these arguments are not enough to change your opinion I will remove this.

More logically I think it would be to have something like `rbd mirror image set $image resync_after_disconnect true`, but I guess it is not worth the effort, at least on this stage...

dillaman replied:
Valid point -- it still might be a little confusing to have a "per-image" config option named with the same prefix as "rbd_mirror"-specific config options. Perhaps ```rbd_resync_to_primary_mirror_after_disconnect``` or something similar?

dillaman commented:
@trociny few minor comments, but otherwise lgtm

trociny commented:
@dillaman Updated. I think I addressed all your comments. The only thing is renaming `rbd_mirror_resync_after_disconnect` config option to `rbd_resync_to_primary_mirror_after_disconnect`. I agree that using `rbd_mirror` prefix is unfortunate here, still I am not very happy with `rbd_resync_to_primary_mirror_after_disconnect` name. I changed it to `rbd_mirroring_resync_after_disconnect` instead. What do you think? If you still like `rbd_resync_to_primary_mirror_after_disconnect` more I will change it.

dillaman commented:
@trociny That's fine with me -- just wanted to ensure it didn't overlap with "rbd_mirror".

dillaman commented:
@trociny rebase required

trociny commented:
@dillaman Testing after rebase I noticed the issue with `ImageReplayer<I>::handle_remote_journal_metadata_updated()`: it might call when ImageReplayer was stopping and shutting journal down, and metadata was null so it crashed.

I have updated handle_remote_journal_metadata_updated to check the current ImageReplayer state and return if it is not running.

Now, retesting this locally. I will let you know about test results.

trociny commented:
@dillaman it passed local tests
